### PR TITLE
add corner pooling

### PR DESCRIPTION
**Pull request by tianyuandu**

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Please add unit tests to ensure the functionality of center pool methods.

File: mmdet/ops/center_pool/__init__.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input = ctx.saved_variables[0]
+        output = center_pool_ext.right_backward(input, grad_output)[0]
+        return output
+
+
+class TopPool(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/ops/center_pool/__init__.py:67]: We may encapsulate `CornerPool` with a `mode` that could handle the need of `TopPool`, `BottomPool`, `LeftPool`, and `RightPool`.

File: tests/test_center_pool.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+def test_center_pool_device_and_dtypes_cpu():
+    """
+    CommandLine:
+        xdoctest -m tests/test_center_pool.py \
+            test_center_pool_device_and_dtypes_cpu
+    """
+    lr_tensor = torch.tensor([[[[0, 0, 0, 0, 0], [2, 1, 3, 0, 2],
+                                [5, 4, 1, 1, 6], [0, 0, 0, 0, 0],
+                                [0, 0, 0, 0, 0]]]])
+    # tb_tensor = torch.tensor([[[[0, 3, 1, 0, 0], [0, 1, 1, 0, 0],
```
Comment (ZwwWayne) [inline tests/test_center_pool.py:19]: Either use or remove these commented code.
Each case need to be covered, including the assert using `pytest.raise(AssertionError)`.

File: mmdet/ops/center_pool/__init__.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    @staticmethod
+    def backward(ctx, grad_output):
+        input = ctx.saved_variables[0]
+        output = center_pool_ext.right_backward(input, grad_output)[0]
+        return output
+
+
+class CenterPool(nn.Module):
+
+    mode_functions = {
```
Comment (ZwwWayne) [inline mmdet/ops/center_pool/__init__.py:69]: mode_functions -> pool_functions

File: mmdet/ops/center_pool/__init__.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input = ctx.saved_variables[0]
+        output = center_pool_ext.right_backward(input, grad_output)[0]
+        return output
+
+
+class CenterPool(nn.Module):
```
Comment (hellock) [inline mmdet/ops/center_pool/__init__.py:67]: Add a docstring.

File: mmdet/ops/center_pool/src/center_pool.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
+#include <torch/torch.h>
```
Comment (hellock) [inline mmdet/ops/center_pool/src/center_pool.cpp:1]: Format cpp code with clang.

File: mmdet/ops/corner_pool/__init__.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
+from torch import nn
```
Comment (hellock) [inline mmdet/ops/corner_pool/__init__.py:1]: Add a `corner_pool.py` and move all implementations to that file.

File: mmdet/ops/corner_pool/src/corner_pool.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
+#include <torch/torch.h>
```
Comment (hellock) [inline mmdet/ops/corner_pool/src/corner_pool.cpp:1]: If this file is modified from other repos, add a reference.

File: mmdet/ops/corner_pool/__init__.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input = ctx.saved_variables[0]
+        output = corner_pool_ext.right_backward(input, grad_output)[0]
+        return output
+
+
+class CornerPool(nn.Module):
+    """ Corner Pooling
```
Comment (hellock) [inline mmdet/ops/corner_pool/__init__.py:68]: No space is needed: `"""Corner Pooling`

File: mmdet/ops/corner_pool/src/corner_pool.cpp (hunk 7 above)
Comment (hellock) [reply mmdet/ops/corner_pool/src/corner_pool.cpp:1]: Although only CPU mode is supported, we may preserve the APIs.

Comment (tianyuandu) [reply mmdet/ops/corner_pool/src/corner_pool.cpp:1]: > Although only CPU mode is supported, we may preserve the APIs.

Should I do any modification to "preserve the APIs"?

**Pull request merged**